### PR TITLE
drivers/periph: renamed random and more

### LINKED ISSUE #3109: periph/random: random_read should return unsigned int
https://github.com/RIOT-OS/RIOT/pull/3101#discussion_r31214572

The documentation for `periph/random.h` says:

>Reads num or less bytes of randomness from the source, will block until random data is available.
>
>Parameters
>    [in]	buf	destination buffer to write the bytes to
>    [in]	num	number of bytes to get from device, only values >0 are valid
>
>Returns
>    the number of bytes written to buf 
>    0 on error 

so we should not need to have a signed int for return type.

### PR DESCRIPTION
**Pull request by haukepetersen**
The random peripheral driver interface needed an overhaul to comply with the other modernized peripheral interfaces:
- renamed it to 'hrng' -> many people complained about the old name...
- removed power on/off functions (as the device should be powered off by default and only powered on while used)
- none of our implementations have any valid error case, so no need for a return value on 'read'

This also fixes #3109

Travis won't be happy, as the path for the `test/pkg_micro-ecc` is still untouched.. .Would somebody mind to tell me the fasted way to change that patch?

### CODE REVIEW

**Comment by gebart** [comment]
How about hwrng instead of hrng? I think hwrng makes it more obvious what this is to the uninitiated reader.

File: tests/pkg_micro-ecc/main.c [hunk 1]
```diff
@@ -61,9 +61,9 @@ int main(void)
     uint8_t l_sig[uECC_BYTES * 2];
 
     /* initialize hardware random number generator */
-    random_init();
+    hrng_init();
     /* power off RNG to save energy */
-    random_poweroff();
+    hrng_poweroff();
```
Comment (gebart) [inline tests/pkg_micro-ecc/main.c]: poweroff was removed, was it not?

**Comment by gebart** [comment]
I agree with the general idea, `random` is too generic and should be changed :+1:

**Comment by kaspar030** [comment]
> I agree with the general idea, random is too generic and should be changed :+1:

:+1:

Or maybe just rng?

**Comment by gebart** [comment]
@kaspar030 doesn't that make it a bit confusing since we have software solutions prefixed with prng?

**Comment by haukepetersen** [comment]
Actually, i would also prefer `hwrng` not that I think about it. @kaspar030, are you okay with that, then I will change it.

**Comment by kaspar030** [comment]
Fine with me!

**Comment by haukepetersen** [comment]
how about this? let me know when I can squash. And also help with adapting this stupid patch file for `pkg/micro-ecc` would be appreciated.

File: boards/mulle/include/periph_conf.h [hunk 2]
```diff
@@ -505,9 +505,8 @@ extern "C"
  * @name Random Number Generator configuration
  * @{
  */
-#define RANDOM_NUMOF            (1U)
-#define RANDOM_CLKEN()          (BITBAND_REG32(SIM->SCGC3, SIM_SCGC3_RNGA_SHIFT) = 1)
-#define RANDOM_CLKDIS()         (BITBAND_REG32(SIM->SCGC3, SIM_SCGC3_RNGA_SHIFT) = 0)
+#define HRNG_CLKEN()        (BITBAND_REG32(SIM->SCGC3, SIM_SCGC3_RNGA_SHIFT) = 1)
```
Comment (kaspar030) [inline boards/mulle/include/periph_conf.h]: HWRNG

File: boards/pba-d-01-kw2x/include/periph_conf.h [hunk 3]
```diff
@@ -325,10 +325,9 @@ extern "C"
  * @name Random Number Generator configuration
  * @{
  */
-#define RANDOM_NUMOF         (1U)
-#define KINETIS_RNGA         RNG
-#define RANDOM_CLKEN()       (SIM->SCGC6 |= (1 << 9))
-#define RANDOM_CLKDIS()      (SIM->SCGC6 &= ~(1 << 9))
+#define KINETIS_RNGA        RNG
+#define HRNG_CLKEN()        (SIM->SCGC6 |= (1 << 9))
```
Comment (kaspar030) [inline boards/pba-d-01-kw2x/include/periph_conf.h]: HWRNG

File: cpu/kinetis_common/hwrng_rnga.c [hunk 4]
```diff
@@ -13,7 +13,7 @@
  * @{
  *
  * @file
- * @brief       Low-level random number generator driver implementation.
+ * @brief       HRNG interface implementation
```
Comment (kaspar030) [inline cpu/kinetis_common/hwrng_rnga.c]: here

File: cpu/kinetis_common/hwrng_rnga.c [hunk 5]
```diff
@@ -22,21 +22,24 @@
...
+    /* nothing to do here */
 }
 
-int random_read(char *buf, unsigned int num)
+void hwrng_read(uint8_t *buf, unsigned int num)
 {
     unsigned int count = 0;
 
+    /* power on and enable the device */
+    HRNG_CLKEN();
```
Comment (kaspar030) [inline cpu/kinetis_common/hwrng_rnga.c]: here

File: cpu/kinetis_common/hwrng_rnga.c [hunk 6]
```diff
@@ -50,31 +53,14 @@ int random_read(char *buf, unsigned int num)
...
-{
+    /* power of the device */
     KINETIS_RNGA->CR = 0;
-    RANDOM_CLKDIS();
-}
-
-/*
-void isr_rng(void)
-{
+    HRNG_CLKDIS();
```
Comment (kaspar030) [inline cpu/kinetis_common/hwrng_rnga.c]: here

File: cpu/kinetis_common/hwrng_rngb.c [hunk 7]
```diff
@@ -13,32 +13,42 @@
...
-    random_poweron();
+    /* nothing to be done here */
 }
 
-int random_read(char *buf, unsigned int num)
+void hwrng_read(uint8_t *buf, unsigned int num)
 {
     unsigned int count = 0;
 
+    HRNG_CLKEN();
```
Comment (kaspar030) [inline cpu/kinetis_common/hwrng_rngb.c]: ...

File: boards/mulle/include/periph_conf.h (hunk 2 above)
Comment (haukepetersen) [inline boards/mulle/include/periph_conf.h]: don't care, its a CPU internally function... But if you insist...

**Comment by kaspar030** [comment]
You forgot the capital HRNG's.

File: boards/pba-d-01-kw2x/include/periph_conf.h (hunk 3 above)
Comment (haukepetersen) [inline boards/pba-d-01-kw2x/include/periph_conf.h]: same

**Comment by gebart** [comment]
@haukepetersen make another patch on top of the 0002-Include-RIOT-Hardware-RNG-interface.patch
make your changes in the checked-out copy of micro-ecc and use git format-patch to create the patch file, then add it as 0003-RIOT-hwrng-API-changes.patch
The two patches can also be merged into one if desired, just use `git diff HEAD~2 > merged.patch`

I'm assuming the actual change is not much more than a search and replace random_ to hwrng_, and removing the random_poweron/off calls?

**Comment by haukepetersen** [comment]
Left them out because they are cpu internal, but I will fix it to make it look nicer...

**Comment by haukepetersen** [comment]
@gebart: yepp, just stome string replace and remove operations. Will do.

**Comment by haukepetersen** [comment]
@kaspar030: fixed.

**Comment by haukepetersen** [comment]
okay, this should be it. If you guys are happy I would squash and give Travis a run at it

**Comment by kaspar030** [comment]
pls squash!

**Comment by haukepetersen** [comment]
squashed

**Comment by gebart** [comment]
the commit messages need updating hrng->hwrng

**Comment by gebart** [comment]
Could you use `format-patch -w` for pkg/micro-ecc/0003-Adapted-to-changed-random-interface.patch to get rid of all the empty lines being stripped of whitespace? it makes the actual patch content less readable.

**Comment by haukepetersen** [comment]
yep, will do both things

**Comment by gebart** [comment]
nucleo-f401 fails:
```
default:nucleo-f401:
Building application "default" for "nucleo-f401" with MCU "stm32f4".
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c: In function 'hwrng_read':
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:34:21: error: 'RCC_AHB2ENR_RNGEN' undeclared (first use in this function)
     RCC->AHB2ENR |= RCC_AHB2ENR_RNGEN;
                     ^
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:34:21: note: each undeclared identifier is reported only once for each function it appears in
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:35:5: error: 'RNG' undeclared (first use in this function)
     RNG->CR = RNG_CR_RNGEN;
     ^
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:35:15: error: 'RNG_CR_RNGEN' undeclared (first use in this function)
     RNG->CR = RNG_CR_RNGEN;
               ^
/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph/hwrng.c:40:28: error: 'RNG_SR_DRDY' undeclared (first use in this function)
         while (!(RNG->SR & RNG_SR_DRDY));
                            ^
make[3]: *** [/home/travis/build/RIOT-OS/RIOT/examples/default/bin/nucleo-f401/periph/hwrng.o] Error 1
make[2]: *** [ALL--/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4/periph] Error 2
make[1]: *** [ALL--/home/travis/build/RIOT-OS/RIOT/cpu/stm32f4] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

I guess it's from the removed `#if RANDOM_NUMOF` at the top of the stm32f4 rng driver?

**Comment by haukepetersen** [comment]
- changed commit messages
- created `micro-ecc` path with `-w` option
- added guard in `cpu/stm32f4/periph/hwrng.c`, as not all F4 variants come with an on-chip RNG. This fixes the issue with the `nucleo-f401`

**Comment by haukepetersen** [comment]
rebased again

**Comment by gebart** [comment]
Travis seem to fail randomly on applying the micro-ecc patches "Makefile already exists in index"
There were also some other random errors in Travis, so I kicked it again.

```
pkg_micro-ecc:stm32f4discovery:
Building application "micro-ecc" for "stm32f4discovery" with MCU "stm32f4".
cd "/home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0001-Add-RIOT-Makefile.patch" || { git am --abort; exit 1; }; cd "/home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0002-Include-RIOT-Hardware-RNG-interface.patch" || { git am --abort; exit 1; }; cd "/home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0003-Adapted-to-changed-random-interface.patch" || { git am --abort; exit 1; };
Applying: Add RIOT Makefile
error: Makefile: already exists in index
Patch failed at 0001 Add RIOT Makefile
The copy of the patch that failed is found in:
   /home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc/.git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
make[1]: *** [/home/travis/build/RIOT-OS/RIOT/pkg/micro-ecc/micro-ecc/Makefile] Error 1
make: *** [/home/travis/build/RIOT-OS/RIOT/tests/pkg_micro-ecc/bin/stm32f4discovery/micro-ecc.a] Error 2
```

**Comment by gebart** [comment]
needs rebase after #4767 was merged.
There's something wrong with the new patch to micro-ecc, I can't apply it on my local checkout.

**Comment by gebart** [comment]
maybe we should just squash the patch for micro-ecc into a single file and be done, what's your opinion @haukepetersen?

**Comment by haukepetersen** [comment]
rebased for now, will look at the patch file once I am in the office. I guess squashing all patches sounds reasonable to me.

**Comment by haukepetersen** [comment]
fuck these patches, I can't get them to work anymore... This is annoying to spend time on this crap

**Comment by gebart** [comment]
On Wed, Feb 10, 2016 at 10:47 AM, Hauke Petersen
<notifications@github.com> wrote:
>
> fuck these patches, I can't get them to work anymore... This is annoying to spend time on this crap

Agreed! let's see if this works... I have attached a fixed patch file
to this email.

**Comment by haukepetersen** [comment]
nope, does still not build...

**Comment by haukepetersen** [comment]
```
"make" -C /home/hauke/dev/riot/RIOT/pkg/micro-ecc
test -d micro-ecc || \
	git clone https://github.com/kmackay/micro-ecc.git /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc && \
		cd /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc && git reset --hard f3d46f0fb77b42535168a17b4b51802beb124a32
HEAD is now at f3d46f0 Fix Thumb-2 inline asm (#15)
cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0001-Add-RIOT-Makefile.patch" || { git am --abort; exit 1; }; cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0002-Include-RIOT-Hardware-RNG-interface.patch" || { git am --abort; exit 1; }; cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0003-Adapted-to-changed-random-interface.patch" || { git am --abort; exit 1; };
previous rebase directory /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc/.git/rebase-apply still exists but mbox given.
make[1]: *** [/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc/Makefile] Error 1
make: *** [/home/hauke/dev/riot/RIOT/tests/pkg_micro-ecc/bin/native/micro-ecc.a] Error 2
```

**Comment by gebart** [comment]
@haukepetersen I probably messed up the patch...

**Comment by haukepetersen** [comment]
sounds familiar :-)

**Comment by gebart** [comment]
@haukepetersen can you try importing these files? https://gist.github.com/gebart/85e68a0382bf6cb1e1cf

**Comment by gebart** [comment]
related: #4788

**Comment by haukepetersen** [comment]
no luck
```
"make" -C /home/hauke/dev/riot/RIOT/pkg/micro-ecc
test -d micro-ecc || \
	git clone https://github.com/kmackay/micro-ecc.git /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc && \
		cd /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc && git reset --hard f3d46f0fb77b42535168a17b4b51802beb124a32
HEAD is now at f3d46f0 Fix Thumb-2 inline asm (#15)
cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0001-Add-RIOT-Makefile.patch" || { git am --abort; exit 1; }; cd "/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc" && git am "../0002-Include-RIOT-Hardware-RNG-interface.patch" || { git am --abort; exit 1; };
Applying: Add RIOT Makefile
Applying: Include RIOT Hardware RNG interface
error: patch failed: uECC.c:322
error: uECC.c: patch does not apply
error: patch failed: uECC.h:57
error: uECC.h: patch does not apply
Patch failed at 0001 Include RIOT Hardware RNG interface
The copy of the patch that failed is found in:
   /home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc/.git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
make[1]: *** [/home/hauke/dev/riot/RIOT/pkg/micro-ecc/micro-ecc/Makefile] Error 1
make: *** [/home/hauke/dev/riot/RIOT/tests/pkg_micro-ecc/bin/airfy-beacon/micro-ecc.a] Error 2
```

**Comment by haukepetersen** [comment]
rebased and pushed those adapted patch files anyway to keep this branch clean...

**Comment by haukepetersen** [comment]
ok, now it works, but it's not pretty. I squashed my changes into the 0002xxx patch, but it only worked without the `-w` option and by re-creating both path files...

**Comment by gebart** [comment]
@haukepetersen did you possibly copy+paste the patch files from the gist above using your editor which strips trailing whitespace? It won't apply if you strip the trailing spaces.
Also, you need to reset the micro-ecc working copy if you are going to try and apply a new patch, preferably rm -rf the whole working copy.

**Comment by haukepetersen** [comment]
yes, my editor actually is removing white spaces, so that might caused trouble... I will try again

**Comment by haukepetersen** [comment]
@gebart alright, that did the trick, now using your proposed patchfiles

**Comment by gebart** [comment]
something is up with the static checks, the cppcheck message got cut off

```
entPtrArg): Assignment of function parameter has no effect outside the function.
```

**Comment by gebart** [comment]
the full message is:
```
cpu/native/periph/hwrng.c:86: style (uselessAssignmentArg): Assignment of function parameter has no effect outside the function.
```

**Comment by haukepetersen** [comment]
fixed, rebase, squashed

**Comment by gebart** [comment]
Travis is green, let's merge this!

**Comment by haukepetersen** [comment]
Awesome! @gebart thanks very much for your help!

**Comment by kaspar030** [comment]
Yeah, nice one!

**Pull request merged**